src/main/io/gps.c

```diff
@@ -843,28 +1373,37 @@ void gpsUpdate(timeUs_t currentTimeUs)
 
     // read out available GPS bytes
     if (gpsPort) {
+        DEBUG_SET(DEBUG_GPS_CONNECTION, 7, serialRxBytesWaiting(gpsPort));
+        static uint8_t wait = 0;
+        static bool isFast = false;
         while (serialRxBytesWaiting(gpsPort)) {
-            if (cmpTimeUs(micros(), currentTimeUs) > GPS_MAX_WAIT_DATA_RX) {
-                // Wait 1ms and come back
-                rescheduleTask(TASK_SELF, TASK_PERIOD_HZ(TASK_GPS_RATE_FAST));
+            int bytes_processed = 0;
+            if ((++bytes_processed % BYTES_MAX) == 0 && cmpTimeUs(micros(), currentTimeUs) > GPS_MAX_WAIT_DATA_RX) {
                 return;
             }
+            wait = 0;
             gpsNewData(serialRead(gpsPort));
+            if (!isFast) {
+                rescheduleTask(TASK_SELF, TASK_PERIOD_HZ(TASK_GPS_RATE_FAST));
+                isFast = true;
+            }
         }
-        // Restore default task rate
-        rescheduleTask(TASK_SELF, TASK_PERIOD_HZ(TASK_GPS_RATE));
-   } else if (GPS_update & GPS_MSP_UPDATE) { // GPS data received via MSP
+
+        if (wait < 1) {
+            wait++;
+        } else if (wait == 1) {
+            wait++;
+            // wait one iteration be sure the buffer is empty, then reset to the slower task interval
+            isFast = false;
+            rescheduleTask(TASK_SELF, TASK_PERIOD_HZ(TASK_GPS_RATE));
+        }
```

Comment:
I'm not sure `wait` and `fast`, and `micros()` are necessary at all.

I would suggest simplifying:

```
int bytes_processed = 0;
        while (serialRxBytesWaiting(gpsPort)) {
            gpsNewData(serialRead(gpsPort));

            if ((++bytes_processed % BYTES_MAX) == 0) {
                rescheduleTask(TASK_SELF, TASK_PERIOD_HZ(TASK_GPS_RATE_FAST));
                return;
            }
        }

        rescheduleTask(TASK_SELF, TASK_PERIOD_HZ(TASK_GPS_RATE));
```

Why is this enough?
- We will read up to 20 bytes, if there is any data pending
- If we care about predictability of timing, checking at every 20 bytes (rather than yielding control back after the first 20 bytes) might yield the following scenario: read 20 bytes, check `micros()`, which shows 29us. We read another 20 bytes, after which we are at 58us, well beyond the limit we wanted ⚠️
- It is more than possible that the `micros()` comparison was always true anyway on most FCs, so the current code always yielded after processing 20 bytes (but does an unnecessary `micros()` computation
- It is actually enough to read 20 bytes and yield the control back to deal with other tasks, at 115200 our reading speed is fast enough to deplete the buffer completely
- If we emptied the buffer, we reschedule to check in normal schedule, no need to do another fast-read cycle if the buffer is empty already
- If there is still data in the buffer, we reschedule with a fast recheck rate and try to empty is fast

Reply:
I spent about 4 hours testing alternatives this evening.
The current code speeds the GPS task up as soon as any data comes in.
This is important, because the GPS task may be running slowly, and only a few characters, out of a large Sat packet, may arrive before the buffer is checked.  In that case, if the timeout (or character count limit) isn't reached, the task will not speed up, and we will wait 100ms before checking the buffer again.  In the meantime, the rest of the incoming packet has 100ms to completely fill the buffer.  
Hence as soon as any characters come in, the task must be sped up, in anticipation of more characters arriving soon after.
And, vice versa, the buffer must be truly empty before slowing down.  If there are some small number of characters in the incoming buffer, below any of the limits, we still need to keep the GPS task running 'fast' so that, no matter few, they won't be waiting another 100ms.
In your code, above, the task won't speed up until there are at least the 'max bytes' number of characters present at the time the buffer is checked.  This causes delays before we deal with the rest of the packet. 
Also it seems that, in the code above, once the buffer is clear, the task will be instructed to run slowly every time the task runs. That seems unnecessary.
I did some tests of simplifications like your proposal, and none worked out much better than the current code.  The current code is limited primarily by the 30us timeout.  At 115200, the timeout is hit when something like 60-70 characters are in the buffer when we first start trying to empty it.  
My AikonF7 still can only run 8k4k despite avoiding the multiple millis() calculations.

Reply:
Thanks for the thorough testing, @ctzsnooze 

> In that case, if the timeout (or character count limit) isn't reached, the task will not speed up, and we will wait 100ms before checking the buffer again. 

It only waits 10ms, and I recommended to change this to 5m worst case (`TASK_GPS_RATE = 200`).



> [from @ledvinap] It's actually 10ms. With that value, the delay shall be reasonable and only about 110 bytes shall end up in buffer. It is strange if something different is happening.

I agree, @ledvinap. If this is not the case, we still have a bug somewhere - I suspect this is the case though and my recommended changes are fine.

> In the meantime, the rest of the incoming packet has 100ms to completely fill the buffer.

10ms at most, which is approx.* 128 bytes at 115200. Reducing to 5ms would yield only approx.* 64 bytes in the buffer, allowing us to lower the buffer sizes and get some memory back (* approx. as due to unfortunate timings these might be blown up by a few bytes).

> Hence as soon as any characters come in, the task must be sped up, in anticipation of more characters arriving soon after.

We _always_ anticipate more bytes to come. 

Keeping the fast rate when the buffer is empty is not be necessary and it's OK to wait some bytes to arrive before we start processing again.

Regardless, the new `TASK_GPS_RATE` is still a good idea.

> And, vice versa, the buffer must be truly empty before slowing down.

That's what my recommendation / original implementation does - once the buffer is empty, reschedule with slower rate.

> If there are some small number of characters in the incoming buffer, below any of the limits, we still need to keep the GPS task running 'fast' so that, no matter few, they won't be waiting another 100ms.

10ms, and recommended 5ms :)
If there are any number of bytes in the buffer, we either read it now (<20), or return in 1ms (according to my recommended changes).

> In your code, above, the task won't speed up until there are at least the 'max bytes' number of characters present at the time the buffer is checked. This causes delays before we deal with the rest of the packet.

In my suggested changes, we either read the full buffer (<20), or if there are still bytes left in the buffer, we reschedule the task to be called again in 1ms. I also suggested reducing the default check-time from 10ms down to 5ms so we don't let the buffer fill up to 128bytes. I don't see any issues with this.

> Also it seems that, in the code above, once the buffer is clear, the task will be instructed to run slowly every time the task runs. That seems unnecessary.

OK, it seems like your main concern is that we don't check the buffer often enough, sometimes waiting 10ms before returning. Mind you, the current code does not guarantee that it won't be the case.

I'm recommending changing the slow-rate to at most 5ms.
At the same time, we need to decide whether the dual-speed is necessary at all. If we prefer to have a default slow rate setting which changes to fast rate if the buffer was not emptied in one go, then my code above does just that.

> I did some tests of simplifications like your proposal, and none worked out much better than the current code. 

Thanks again for testing.

It should not work *much better* - or any better in most cases :)
It should just work, with the benefits outlined above:
- simpler code (this is way more important than it might seem at first)
- smaller buffer requirement (reduce the buffers down to 128 again)
- more predictability due to more frequent yields
- more chance not to blow the 30us limit


> The current code is limited primarily by the 30us timeout. At 115200, the timeout is hit when something like 60-70 characters are in the buffer when we first start trying to empty it.

This is the exact reason we want to yield more often - see my original post re why reading in batches can, and does, the blow the timeouts ("If we care about predictability of timing, checking at every 20 bytes...")

> My AikonF7 still can only run 8k4k despite avoiding the multiple millis() calculations.

`micros()` :) Don't forget about poor F4 folks out there, the timing and memory changes would benefit them more.

The code-simplicity changes would benefit everyone when we avoid non-trivial bugs

Reply:
@adam-ah
Thanks for your comments, above. What you say does make a lot of sense.
My mistake to say the wait was 100ms, I meant to say 10ms.  

My tests agree with your observations that at 115200 the max incoming bytes, the max expected incoming serial buffer at 10hz definitely seems to be around 100.
Here is an image of the current code, where we see several seconds of data, with the pink 'NotUsed' trace showing the size of the serial port buffer over time.
Note this particular peak at 98bytes; it's about the biggest in this brief log.
The wider pink blocks are the sequential acquisition of a sat packet.  For them, and the better timed Nav packets, they buffer is around 23-24 bytes.
![Screen Shot 2023-08-04 at 12 53 14(https://github.com/betaflight/betaflight/assets/11737748/1e41dfbb-f62a-49f2-8eb0-8e1ae35d54ca)

Here we look at 1s of data and can see more detail about the interval between task runs (climbing yellow staircase, with 'normal' intervals of 10ms).
![Screen Shot 2023-08-04 at 12 19 11](https://github.com/betaflight/betaflight/assets/11737748/59e2d5a5-edd1-41e4-85d7-dbb5ad8ec4ff)

Unfortunately, the task scheduler currently has issues (that I cannot explain) with my F7 at 8k8k.  The normal task rate becomes 50hz and the fast task at 250hz, or half speed for both.  The actual interval then becomes 20ms, and the 115200 baud buffer typically peaks around 200-230 bytes.  This shows the intervals, with the same time scale as the previous image.

![Screen Shot 2023-08-04 at 12 14 09](https://github.com/betaflight/betaflight/assets/11737748/ead211b4-569d-4ec6-8ffa-d77f3f5d401f)

And here is a 1s period showing a large spike to 231 on the input buffer while at 8k4k
![Screen Shot 2023-08-04 at 13 01 54](https://github.com/betaflight/betaflight/assets/11737748/bfa1fd22-c3e3-43f7-ada7-3f238e09964d)

I agree that simple code is generally preferred, and your suggestion is indeed simple.  

The main concern isn't the virtue of simple code.  It is the downside of having the task over-run.  The scheduler is co-operative, and if we do find the intervals become larger as shown above, and we do get a lot of bytes to handle, which, unless constrained, will cause a bigger scheduler over-run the next time around.  That's why we limit the number of microseconds allowed in the while loop; we need to end this task to allow others to complete.

What is curious about the 8k8k log is that even when the incoming buffer is zero, we still find the scheduler is delaying the gps task.  When that buffer is zero, and we are in 'receiving data' mode, there is just about nothing to do.  So we should be completed in very short time.  So it's a bit strange to see this problem.

Scheduler related issues are complex. I do not understand them. @SteveCEvans is the expert; I've asked his advice and investigation.  In the meantime I think it best to perform tests of the kind I have done, while experimenting with different options.  I did try your solution, but we would need to log the time taken in microseconds, without the limiter, to see just exactly how long some things could take.

Reply:
PS if you look very closely at the last image, there are some times when we see a speed up in the steps on the yellow trace, when we acquire a Nav packet, and other times we don't.  If the Nav packet hasn't hit the time limit, the code doesn't return, and the yellow trace updates promptly because the code reaches the point where that debug is updated (down the bottom of `gpsUpdate()`).  With larger amounts of data in the incoming buffer, the return is triggered, and although we see steps in the buffer as it empties, we do not see steps in the yellow trace. That's how I can tell whether the time-out is being hit.  
In the image above, it is being hit not infrequently, by any incoming buffer size above about 70-80 bytes, since bytes can be added while in the while loop.  If the incoming buffer size is above that value, we hit the timeout.  
From this i know that the timeout in the real-world of this FC is not hard to hit at 8k4k with this scheduler limitation.

Reply:
> The scheduler is co-operative, and if we do find the intervals become larger as shown above, and we do get a lot of bytes to handle, which, unless constrained, will cause a bigger scheduler over-run the next time around. That's why we limit the number of microseconds allowed in the while loop; we need to end this task to allow others to complete.

I'm not sure I'm following. My recommended change, on average, would yield control back the scheduler more often' and, on average, it would less likely to breach the 30us limit than the current code. 

Checking the clock after every 20 bytes and continuing (instead of yielding) is likely to breach the 30us limit, by design:
- on average, would yield at 45us
- worst case it would yield at 60us
- best case in less than 30us (when the buff size was <20 bytes)

> Unfortunately, the task scheduler currently has issues (that I cannot explain) with my F7 at 8k8k. The normal task rate becomes 50hz and the fast task at 250hz, or half speed for both.

if this is the case and we can't rely on the scheduler, then are in all sorts of trouble, outside of this specific problem here. @SteveCEvans, any thoughts on this?